apps/passport-server/test/ratelimit.spec.ts

```diff
@@ -0,0 +1,173 @@
+import { expect } from "chai";
+import "mocha";
+import { step } from "mocha-steps";
+import { Pool } from "postgres-pool";
+import { getDB } from "../src/database/postgresPool";
+import {
+  checkRateLimit,
+  clearExpiredRateLimits
+} from "../src/database/queries/rateLimit";
+import { sqlQuery } from "../src/database/sqlQuery";
+import { overrideEnvironment, testingEnv } from "./util/env";
+
+describe("generic rate-limiting features", function () {
+  this.timeout(15_000);
+
+  let db: Pool;
+
+  this.beforeAll(async () => {
+    await overrideEnvironment(testingEnv);
+    db = await getDB();
+  });
+
+  this.afterAll(async () => {
+    await db.end();
+  });
+
+  step("database should initialize", async function () {
+    expect(db).to.not.eq(null);
+  });
+
+  /**
+   * At the time of writing, 10 token check attempts are allowed per-hour.
+   */
+  step(
+    "password reset should not be rate-limited at first attempt",
+    async function () {
+      const result = await checkRateLimit(
+        db,
+        "CHECK_EMAIL_TOKEN",
+        "test@example.com"
+      );
+
+      expect(result).to.be.true;
+    }
+  );
+
+  step(
+    "password reset should be rate-limited after >10 attempts",
+    async function () {
+      for (let i = 0; i < 9; i++) {
+        const result = await checkRateLimit(
+          db,
+          "CHECK_EMAIL_TOKEN",
+          "test@example.com"
+        );
+
+        expect(result).to.be.true;
+      }
+
+      const exceededLimitResult = await checkRateLimit(
+        db,
+        "CHECK_EMAIL_TOKEN",
+        "test@example.com"
+      );
+
+      expect(exceededLimitResult).to.be.false;
+    }
+  );
+
+  step(
+    "password reset should not be rate-limited for a different unique event ID such as e-mail address",
+    async function () {
+      const result = await checkRateLimit(
+        db,
+        "CHECK_EMAIL_TOKEN",
+        "differentemail@example.com"
+      );
+
+      expect(result).to.be.true;
+    }
+  );
+
+  step(
+    "earlier limited event ID should still be rate-limited",
+    async function () {
+      const exceededLimitResult = await checkRateLimit(
+        db,
+        "CHECK_EMAIL_TOKEN",
+        "test@example.com"
+      );
+
+      expect(exceededLimitResult).to.be.false;
+    }
+  );
+
+  step(
+    "after time has elapsed, more attempts should be allowed",
+    async function () {
+      // This simulates the case where the "bucket" of available actions has
+      // not been re-filled for 30 minutes. Since 10 such actions are allowed
+      // per hour, a 30-minute wait should allow 5 new attempts.
+      await sqlQuery(
+        db,
+        "UPDATE rate_limit_buckets SET last_refill = NOW() - INTERVAL '30 MINUTES'"
+      );
```

Comment:
thought: if `now` was able to be passed in via typescript, you would not need to do this via sql query, you would be able to fast forward time via `MockDate`